pl_examples/domain_templates/imagenet.py

```diff
@@ -86,53 +80,56 @@ def __init__(
         loss_train = F.cross_entropy(output, target)
-        acc1, acc5 = self.__accuracy(output, target, topk=(1, 5))
-        self.log("train_loss", loss_train, on_step=True, on_epoch=True, logger=True)
-        self.log("train_acc1", acc1, on_step=True, prog_bar=True, on_epoch=True, logger=True)
-        self.log("train_acc5", acc5, on_step=True, on_epoch=True, logger=True)
+        self.log("train_loss", loss_train)
+        # update metrics
+        self.train_acc1(output, target)
+        self.train_acc5(output, target)
+        self.log("train_acc1", self.train_acc1, prog_bar=True)
+        self.log("train_acc5", self.train_acc5, prog_bar=True)
         return loss_train
 
     def eval_step(self, batch, batch_idx, prefix: str):
         images, target = batch
         output = self(images)
         loss_val = F.cross_entropy(output, target)
-        acc1, acc5 = self.__accuracy(output, target, topk=(1, 5))
-        self.log(f"{prefix}_loss", loss_val, on_step=True, on_epoch=True)
-        self.log(f"{prefix}_acc1", acc1, on_step=True, prog_bar=True, on_epoch=True)
-        self.log(f"{prefix}_acc5", acc5, on_step=True, on_epoch=True)
+        self.log(f"{prefix}_loss", loss_val)
+        # update metrics
+        self.eval_acc1(output, target)
+        self.eval_acc5(output, target)
+        self.log(f"{prefix}_acc1", self.eval_acc1, prog_bar=True)
+        self.log(f"{prefix}_acc5", self.eval_acc5, prog_bar=True)
+        return loss_val
 
     def validation_step(self, batch, batch_idx):
         return self.eval_step(batch, batch_idx, "val")
 
-    @staticmethod
-    def __accuracy(output, target, topk=(1,)):
-        """Computes the accuracy over the k top predictions for the specified values of k."""
-        with torch.no_grad():
-            maxk = max(topk)
-            batch_size = target.size(0)
-
-            _, pred = output.topk(maxk, 1, True, True)
-            pred = pred.t()
-            correct = pred.eq(target.view(1, -1).expand_as(pred))
-
-            res = []
-            for k in topk:
-                correct_k = correct[:k].reshape(-1).float().sum(0, keepdim=True)
-                res.append(correct_k.mul_(100.0 / batch_size))
-            return res
+    def test_step(self, batch, batch_idx):
+        return self.eval_step(batch, batch_idx, "test")
 
     def configure_optimizers(self):
         optimizer = optim.SGD(self.parameters(), lr=self.lr, momentum=self.momentum, weight_decay=self.weight_decay)
         scheduler = lr_scheduler.LambdaLR(optimizer, lambda epoch: 0.1 ** (epoch // 30))
         return [optimizer], [scheduler]
 
+    def setup(self, stage=None):
+        if isinstance(self.trainer.strategy, ParallelStrategy):
+            # When using a single GPU per process and per `DistributedDataParallel`, we need to divide the batch size
+            # ourselves based on the total number of GPUs we have
+            self.batch_size = int(self.batch_size / max(1, self.trainer.devices))
```

Comment:
maybe use num_processes?
I think if someone sets devices=-1, it stays as it is.. and here max(1, -1) won't split the batch_size as mentioned.
same for workers.